pkg/sabakan-state-setter/controller.go

```diff
@@ -199,27 +202,105 @@ func (c *Controller) runOnce(ctx context.Context) error {
 		log.Warn("failed to get sabakan machines", map[string]interface{}{
 			log.FnError: err.Error(),
 		})
-		// lint:ignore nilerr  RunPeriodically tries this again.
+		// lint:ignore nilerr  Run tries this again.
 		return nil
 	}
 	if len(machines) == 0 {
 		log.Warn("no machines found", nil)
 		return nil
 	}
 
+	// Check machine types
+	for _, m := range machines {
+		if _, ok := c.machineTypes[m.Type]; !ok {
+			log.Warn("unknown machine type", map[string]interface{}{
+				"serial":       m.Serial,
+				"machine_type": m.Type,
+			})
+		}
+	}
+
+	// Do machines health check
+	newStateMap := c.machineHealthCheck(ctx, machines)
+
+	// Do machines retire
+	// T.B.D.
+
+	now := time.Now()
+	for _, m := range machines {
+		newState, ok := newStateMap[m.Serial]
+		if !ok || m.State == newState {
+			c.ClearUnhealthy(m)
+			continue
+		}
+		if newState == sabakan.StateUnhealthy {
+			// Wait for the GracePeriod before changing the machine state to unhealthy.
+			if ok := c.RegisterUnhealthy(m, now); !ok {
+				continue
+			}
+		}
+
+		err := c.sabakanClient.UpdateSabakanState(ctx, m.Serial, newState)
+		if err != nil {
+			switch e := err.(type) {
+			case *gqlerror.Error:
+				// In the case of an invalid state transition, the log may continue to be output.
+				// So the log is not output.
+				if eType, ok := e.Extensions["type"]; ok && eType == gqlsabakan.ErrInvalidStateTransition {
+					continue
+				}
+				log.Warn("gql error occurred when set state", map[string]interface{}{
+					log.FnError: err.Error(),
+					"serial":    m.Serial,
+					"ipv4":      m.IPv4Addr,
+				})
+			default:
+				log.Warn("error occurred when set state", map[string]interface{}{
+					log.FnError: err.Error(),
+					"serial":    m.Serial,
+					"ipv4":      m.IPv4Addr,
+				})
+			}
+		} else {
+			log.Info("change state", map[string]interface{}{
+				"serial": m.Serial,
+				"ipv4":   m.IPv4Addr,
+				"state":  newState,
+			})
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) machineHealthCheck(ctx context.Context, machines []*machine) map[string]sabakan.MachineState {
+	// Get serf status
 	serfStatus, err := c.serfClient.GetSerfStatus()
 	if err != nil {
 		log.Warn("failed to get serf members", map[string]interface{}{
 			log.FnError: err.Error(),
 		})
-		// lint:ignore nilerr  RunPeriodically tries this again.
-		return nil
+		return map[string]sabakan.MachineState{}
```

Comment:
The returned map will not be modified, so we don't need to make an empty map here.

```suggestion
		return nil
```